popde/adaptive_kde.py

```diff
@@ -0,0 +1,172 @@
+import numpy as np
+import scipy
+from density_estimate import VariableBwKDEPy 
+from scipy.stats import gmean
+
+class AdaptiveBwKDE(VariableBwKDEPy):
+    """
+    General adaptive Kernel Density Estimation (KDE) using KDEpy
+    with variable bandwidth per point
+    self.alpha  = alpha an additional parameter from B. Wang and X. Wang, 
+    2007, DOI: 10.1214/154957804100000000.
+
+    Example:
+    # Example of 3D KDE and plot for verification
+    mean1, sigma1 = 0.0, 1.0
+    mean2, sigma2 = 0.0, 1.0
+    mean3, sigma3 = 0.0, 1.0
+    n_samples = 1000
+    rndgen = np.random.RandomState(seed=1)
+    sample1 = rndgen.normal(mean1, sigma1, size=n_samples)
+    sample2 = rndgen.normal(mean2, sigma2, size=n_samples)
+    sample3 = rndgen.normal(mean3, sigma3, size=n_samples)
+    sample = np.column_stack((sample1, sample2, sample3)) # shape is (n_points, n_features)
+    # Create and fit the adaptive KDE 
+    kde = AdaptiveBwKDE(sample, dim_names=['x', 'y', 'z'], alpha=0.5, input_transf=None)
+    # Generate grid for plotting
+    xgrid = np.linspace(sample1.min(), sample1.max(), 100)
+    ygrid = np.linspace(sample2.min(), sample2.max(), 100)
+    zgrid = np.linspace(sample3.min(), sample3.max(), 100)
+    XX, YY, ZZ = np.meshgrid(xgrid, ygrid, zgrid)
+    eval_pts = np.column_stack((XX.flatten(), YY.flatten(), ZZ.flatten()))
+
+    # Evaluate the KDE at the grid points
+    density_values = kde.evaluate(eval_pts)
+    density_values = density_values.reshape(XX.shape)
+    print(density_values)
+    """
+    def __init__(self, data, backend='KDEpy', bandwidth=1., dim_names=None, alpha=0.5,  input_transf=None, stdize=False, rescale=None):
+        self.alpha = alpha
+        self.global_bandwidth = bandwidth
+
+        super().__init__(data, input_transf, stdize, rescale,
+                         backend, bandwidth, dim_names) 
+
+        #Adaptive bandwidth: compute pilot kdevals
+        self.pilot_values = self.evaluate(self.kde_data)
+        #Calculate per-point bandwidths to  re-assign self.bandwidth and re-initialize KDE
+        self.set_per_point_bandwidth(self.pilot_values)
+
+    def _local_bandwidth_factor(self, kde_values):
+        """
+        Calculate local bandwidth factor using expression in
+        B. Wang and X. Wang, 2007, DOI: 10.1214/154957804100000000.
+
+        Parameters:
+        -----------
+        kde_values : array-like
+            The KDE values at the data point positions.
+
+        Returns:
+        --------
+        loc_bw_factor : array-like
+           local bandwidth factor for adaptive 
+           bandwidth.
+        """
+        #geometric mean of kde values using scipy
```

Comment:
```suggestion
        # Geometric mean of pilot kde values
```
Since gmean is clearly imported from scipy, no need to mention it here

Reply:
done